phpBB/phpbb/ban/type/ip.php

```diff
@@ -0,0 +1,550 @@
+
+			$start_ip_long = ip2long($start_ip);
+			$end_ip_long = ip2long($end_ip) + 1;
+
+			$number_ips = $end_ip_long - $start_ip_long + 1;
+			$number_ips_bin = str_pad(decbin($number_ips), 32, '0', STR_PAD_LEFT);
+
+			if ($blocks_start_ip[0] === $blocks_end_ip[0])
+			{
+				for ($i = strlen($number_ips_bin); $i > 0; $i--)
+				{
+					if ($number_ips_bin[$i - 1])
+					{
+						$cidr_ranges[] = long2ip($start_ip_long) . '/' . $i;
+						$start_ip_long += (1 << 32 - $i);
+
+						if ($start_ip_long === $end_ip_long + 1)
+						{
+							break;
+						}
+					}
+				}
+			}
+			else
+			{
+				for ($i = 1, $size = strlen($number_ips_bin); $i <= $size; $i++)
+				{
+					if ($number_ips_bin[$i - 1])
+					{
+						$cidr_ranges[] = long2ip($start_ip_long) . '/' . $i;
+						$start_ip_long += (1 << 32 - $i);
+
+						if ($start_ip_long === $end_ip_long + 1)
+						{
+							break;
+						}
+					}
+				}
+			}
+		}
+		else
+		{
+			// @TODO: throw ex
+			return false;
+		}
+
+		return $cidr_ranges;
+	}
+
+	/**
+	 * Adds a subnet to an IPv6 address in its unpacked form
+	 *
+	 * @param array		$bytes_ip	The IPv6 address in its unpackd form (format: n*)
+	 * @param int		$subnet		The subnet which should be added
+	 *
+	 * @return array	Returns the IPv6 address with the added subnet in its unpacked form
+	 */
+	protected function ipv6_add_subnet(array $bytes_ip, $subnet)
+	{
+		if (empty($bytes_ip))
+		{
+			// @TODO: throw ex
+			return false;
+		}
+		if ($subnet < 1 || $subnet > 128)
+		{
+			// @TODO: throw ex
+			return false;
+		}
+
+		$block = (int) ceil($subnet / 16);
+		$position = (128 - $subnet) % 16;
+
+		$bytes_ip[$block] += 1 << $position;
+		// If we got a carry after adding 2^x to the block, we remove it and add one to the next block
+		if (($bytes_ip[$block] & 0x10000) === 0x10000)
+		{
+			$bytes_ip[$block] &= 0xFFFF;
+			for ($i = $block - 1; $i >= 0; $i--)
+			{
+				$bytes_ip[$i] += 1;
+				// ... and so on...
+				if (($bytes_ip[$i] & 0x10000) !== 0x10000)
+				{
+					break;
+				}
+				$bytes_ip[$i] &= 0xFFFF;
+			}
+		}
+
+		return $bytes_ip;
+	}
+
+	/**
+	 * Gets the number of IPs between two IPv6 addresses in their unpacked form (format: n*)
+	 *
+	 * @param array 	$bytes_ip1	The first IPv6 address
+	 * @param array 	$bytes_ip2	The second IPv6 address
+	 *
+	 * @return array	The difference in the same format
+	 */
+	protected function ipv6_count_ips(array $bytes_ip1, array $bytes_ip2)
+	{
+		$swap = false;
+		for ($i = 1; $i <= 8; $i++)
+		{
+			if ($bytes_ip1[$i] > $bytes_ip2[$i])
+			{
+				$swap = true;
+				break;
+			}
+		}
+		if ($swap)
+		{
+			$temp = $bytes_ip1;
+			$bytes_ip1 = $bytes_ip2;
+			$bytes_ip2 = $temp;
+		}
+
+		// We will overwrite all blocks anyway, we just need to keep the order
+		$diff_blocks = $bytes_ip1;
+		// Set carry to -1 to add 1 to the overall result
+		$carry = -1;
+
+		for ($i = 8; $i >= 1; $i--)
+		{
+			$diff = $bytes_ip2[$i] - $bytes_ip1{$i} - $carry;
+			$carry = ($diff < 0) ? 1 : 0;
```

Comment:
When `$bytes_ip2[8] = 0xffff; $bytes_ip1[8] = 0;`, this won't work. 

I know, it is my fault. ;)